### PR TITLE
WIP: Initial config for waf filter

### PR DESCRIPTION
**Pull request by kprakasam**
https://github.com/envoyproxy/envoy/issues/7918
Signed-off-by: Prakasam Kannan <kprakasam@me.com>

Commit Message: Initial WAF filter config for review.
Additional Description: https://github.com/envoyproxy/envoy/issues/7918
Risk Level:
Testing:
Docs Changes:
Release Notes:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Deprecated:]

### CODE REVIEW

**Comment by dio** [comment]
cc. @arifsetiawan

File: api/envoy/extensions/filters/http/waf/v3/waf.proto [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+option java_package = "io.envoyproxy.envoy.extensions.filters.http.waf.v3";
+option java_outer_classname = "WAFProto";
+option java_multiple_files = true;
+option (udpa.annotations.file_status).package_version_status = ACTIVE;
+
+// [#protodoc-title: WAF]
+// WAF :ref:`configuration overview <config_http_filters_waf_tagging>`.
+// [#extension: envoy.filters.http.waf]
+
+message WAF {
```
Comment (htuch) [inline api/envoy/extensions/filters/http/waf/v3/waf.proto:18]: Nit: `Waf` (we don't do all caps message names), e.g. we have `Tcp`.

File: api/envoy/extensions/filters/http/waf/v3/waf.proto [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+
+ message RuleMatchCB {
+   string name = 1;
+   google.protobuf.Any config = 2;
+ }
+
+ // Watcher that inspect the HTTP request and recommend one or more Action
+ message Watcher {
+   string name = 1;
+   google.protobuf.Any config = 2;
```
Comment (htuch) [inline api/envoy/extensions/filters/http/waf/v3/waf.proto:144]: Please use `TypedExtensionConfig` for all these pluggable actions.

File: api/envoy/extensions/filters/http/waf/v3/waf.proto [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+
+    string path = 3;
+
+    string host = 4 [(validate.rules).string.min_len = 0, (validate.rules).string.hostname = true];
+
+    // default is https
+    string scheme = 5 [(validate.rules).string = {in: ["http", "https"]}];
+  }
+
+  message ExternalService {
```
Comment (htuch) [inline api/envoy/extensions/filters/http/waf/v3/waf.proto:217]: What is this for? Can you document?

File: api/envoy/extensions/filters/http/waf/v3/waf.proto [hunk 4]
```diff
@@ -0,0 +1,223 @@
...
+
+      string cluster_redirect = 5;
+
+      HttpRedirect http_redirect = 6;
+
+      ExternalService external_service = 6;
+    }
+  }
+
+  message HttpRedirect {
```
Comment (htuch) [inline api/envoy/extensions/filters/http/waf/v3/waf.proto:203]: Please order this before `Action`. Also, should we reuse existing HTTP redirect messages from `RouteConfiguration` here?

File: api/envoy/extensions/filters/http/waf/v3/waf.proto [hunk 5]
```diff
@@ -0,0 +1,223 @@
...
+  message PolicyGroup {
+    // Name of the Policy Group.
+    string name = 1 [(validate.rules).string.pattern = "^[a-zA-Z0-9]([-a-zA-Z0-9]*[a-zA-Z0-9])?$"];
+    PolicySet policies = 2;
+    RuleMatchCB callback = 3;
+  }
+
+  repeated PolicyGroup policy_groups = 1[(validate.rules).repeated .min_items = 1];
+  Executor executor = 2;
+ }
```
Comment (htuch) [inline api/envoy/extensions/filters/http/waf/v3/waf.proto:113]: Can you reformat this file with `fix_format`? I'm having a hard time grouping messages with indents.

File: api/envoy/extensions/filters/http/waf/v3/waf.proto [hunk 6]
```diff
@@ -0,0 +1,223 @@
...
+
+ // Rule defines a match condition and an associated action.
+ message Policy {
+  string name = 1 (validate.rules).string.pattern = "^[a-zA-Z0-9]([-a-zA-Z0-9]*[a-zA-Z0-9])?$"];
+  Match match = 2;
+  Watcher watcher = 3;
```
Comment (htuch) [inline api/envoy/extensions/filters/http/waf/v3/waf.proto:119]: I'm having a hard time trying to distinguish match and watcher in my head. Can you write a sections in `docs/root/configuration` that explains in detail the model for WAF and how these differ?

File: api/envoy/extensions/filters/http/waf/v3/waf.proto [hunk 7]
```diff
@@ -0,0 +1,223 @@
...
+ }
+
+ // RuleMatchCB that will decide if the rule engine should
+ // 1. Continue evaluating remaining rules that are bound to this callback
+ // 2. Possibly apply request header mutations that Watchers may depend on
+ // 3. Pass along the actions recommended by watchers to others
+
+ message RuleMatchCB {
+   string name = 1;
+   google.protobuf.Any config = 2;
```
Comment (htuch) [inline api/envoy/extensions/filters/http/waf/v3/waf.proto:138]: I'm not sure the notion of "callback" makes a ton of sense at the API level. Do you mean to say `MatchExtension`?

File: api/envoy/extensions/filters/http/waf/v3/waf.proto [hunk 8]
```diff
@@ -0,0 +1,223 @@
+syntax = "proto3";
+
+package envoy.extensions.filters.http.waf.v3;
```
Comment (mattklein123) [inline api/envoy/extensions/filters/http/waf/v3/waf.proto:3]: I would suggest we do v3alpha here for now until we get a bit of experience with the filter. Then we can upgrade it to v3 later.

File: api/envoy/extensions/filters/http/waf/v3/waf.proto [hunk 9]
```diff
@@ -0,0 +1,223 @@
...
+//    policies:
+//    - name: block-admin-pages
+//      match:
+//        header:
+//          name: ":path"
+//          prefix: "/admin"
+//      watcher:
+//        - name: envoy.watcher
+//          action:
+//            reset: truek
```
Comment (mattklein123) [inline api/envoy/extensions/filters/http/waf/v3/waf.proto:42]: typo

File: api/envoy/extensions/filters/http/waf/v3/waf.proto [hunk 10]
```diff
@@ -0,0 +1,223 @@
...
+//    - name: ofac-white-list
+//      match:
+//        sourceIp:
```
Comment (mattklein123) [inline api/envoy/extensions/filters/http/waf/v3/waf.proto:45]: nit: use snake_case for all examples. Same elsewhere.

**Comment by kprakasam** [comment]
@mattklein123 & @htuch I saw your comments, I primarily work on these during weekend please post your comments and suggested I'll respond to them over the weekend. Thanks for the appreciation and support.

**Comment by kprakasam** [comment]
> Thanks for getting this going! I'm trying to grok the basic match/watcher/action model to begin with, would you be able to provide some insight into why we have three things here rather than just match/action?

@mattklein123 & @htuch:

I was also ambivalent about having 4 different extension mechanism (Action, RuleMatchCallback, Watchers and Executor), I based the design on the following premise.

1. Generally an action is either terminal (drop, reset, redirect request) or non terminal (add, modify, delete headers or body) at the time of policy definition.
2. Sometimes an action are neither terminal or non terminal at the time of definition policy but upon execution yields another action which is terminal or non terminal (External service action which invokes an external service let's day a Bot detector which may in turn return either a terminal or a non terminal action)
3. Generally upon first policy match, WAF will execute the associated action and stop further policy evaluation.  
4. Sometimes it is desirable to continue policy evaluation when the action is non terminal (add, modify, remove header)

Current design provides for addressing all combination of the above rules through custom implementation of Action, RuleMatchCallback, Watchers and Executor but complicates the design.

Upon further reflection, I wonder if we added an attribute to the policy that specifies whether the policy evaluation should continue even after the match, and the Action to return a boolean that would essentially say if it is safe to evaluate additional policies and execute associated action (this is necessary since some actions become terminal or otherwise only upon execution), we should be able to address most of the above combinations of rules if not all. 

This will simplify the design a great deal and we can certainly us iterate and improve the design and implementation in the future. Please let me know at very high level if you want to continue with this approach so I can update the design and Protobuf accordingly and address other related issues.

**Comment by htuch** [comment]
Sure, that seems a bit simpler, you could even have `TerminalAction` and `NonTerminalAction` types if this is static, but it would need to be a return if dynamic. It would be interesting to see a comparison with other WAF rule systems. How complete/expressive is this vs. state-of-the-art?

**Comment by rshriram** [comment]
drive by q.. Functionally, what is in there that cannot be accomplished by the existing RBAC filter in envoy, or even just basic route matching that only allows the matched routes to work while denying the rest? Is there a dynamic rule loading component at play here that is missing in other filter types like rbac? If so, can it be accomplished by providing per-filter chain reload ability or heck, even a simple grpc call out in the existing rbac filter that periodically reloads rbac rules from some central server.

**Comment by mattklein123** [comment]
> drive by q.. Functionally, what is in there that cannot be accomplished by the existing RBAC filter in envoy, or even just basic route matching that only allows the matched routes to work while denying the rest? Is there a dynamic rule loading component at play here that is missing in other filter types like rbac? If so, can it be accomplished by providing per-filter chain reload ability or heck, even a simple grpc call out in the existing rbac filter that periodically reloads rbac rules from some central server.

@rshriram raises a good point which is whether we should build a new filter or just built actions into the existing RBAC filter. See https://github.com/envoyproxy/envoy/issues/11594 for an example of this. @davidraskin @liminw

**Comment by kprakasam** [comment]
That’s another way to do it, since we are proposing now to simplify the design to extend functionality by custom actions. Even in the original proposal we intended to use the same matching engine.

**Comment by liminw** [comment]
> > drive by q.. Functionally, what is in there that cannot be accomplished by the existing RBAC filter in envoy, or even just basic route matching that only allows the matched routes to work while denying the rest? Is there a dynamic rule loading component at play here that is missing in other filter types like rbac? If so, can it be accomplished by providing per-filter chain reload ability or heck, even a simple grpc call out in the existing rbac filter that periodically reloads rbac rules from some central server.
> 
> @rshriram raises a good point which is whether we should build a new filter or just built actions into the existing RBAC filter. See #11594 for an example of this. @davidraskin @liminw

+1. My preference is also to extend actions in the existing RBAC filter.

**Comment by mattklein123** [comment]
> +1. My preference is also to extend actions in the existing RBAC filter.

Agreed, let's do that. We can add more actions as needed. @kprakasam can you take a look at https://github.com/envoyproxy/envoy/issues/11594 as inspiration and propose some additional config there for new actions?

**Comment by kprakasam** [comment]
Sure

**Comment by kprakasam** [comment]
Initial Design Doc - v2

https://docs.google.com/document/d/1PYw3BnqIbX93qv3e-jbPY0lv_alJAQlv9ToVI5OKa5s/edit?usp=sharing

**Comment by htuch** [comment]
@kprakasam Thanks.. can you make the doc world commentable?

**Comment by kprakasam** [comment]
I should be editable for any one with the link I posted. Should I do anything else?

**Comment by htuch** [comment]
It doesn't need to be editable, just commentable. It still appears to require a request for "edit access" to comment.

**Comment by mattklein123** [comment]
+1 please add world comment access.

**Comment by kprakasam** [comment]
Commentable link: https://docs.google.com/document/d/1PYw3BnqIbX93qv3e-jbPY0lv_alJAQlv9ToVI5OKa5s/edit?usp=sharing

**Comment by kprakasam** [comment]
@mattklein123, @htuch, et al., I have simplified the config and provided examples of how RBAC, IP tagging, reset, drop, redirect, etc. would be configured as an action. I'd appreciate if you guys can provide feedback on if we want to continue down this path in terms of design.

**Comment by htuch** [comment]
@kprakasam I'd like to know how this and https://github.com/envoyproxy/envoy/issues/11832 relate. Tagging @yangminzhu. I think they are somewhat related, since while #11832 is about access control, there's a common match aspect, and both proposals are about trying to unify interfaces across similar filers, e.g. ext_authz, RBAC.

**Comment by kprakasam** [comment]
> @kprakasam I'd like to know how this and #11832 relate. Tagging @yangminzhu. I think they are somewhat related, since while #11832 is about access control, there's a common match aspect, and both proposals are about trying to unify interfaces across similar filers, e.g. ext_authz, RBAC.

They are very related while  #11832 just try to create shared matching engine, this one will unify the matching engine and provide ability to execute any Action upon matching condition. ext_authz and RBAC will become an action. Please see the example configurations[here](https://github.com/envoyproxy/envoy/pull/11386/files#diff-621244804018f0ecc6077a02e57f880bR117) which shows how RBAC, IP_TAGGING, EXT_AUTH will be configured as an action.

**Comment by htuch** [comment]
@kprakasam @yangminzhu can you folks sync up and see what can be done to reconcile proposals? It would be helpful to have a single proposal or a clear delineation between them to review. Thanks :)

**Comment by yangminzhu** [comment]
We should have a single matching API shared by many different filters (ext_authz, RBAC, tap and possibly this new waf). @kprakasam is there a reason to introduce a new matching API in the waf filter? does the unified matching API (https://shorturl.at/gkrM3) satisfy your need here? (Please let me know if it doesn't and we can improve it to support)

@htuch my understanding is that this PR is doing something more than the matching API part that I proposed in https://github.com/envoyproxy/envoy/issues/11832, the unified matching API should be part of this new waf filter as well.

**Comment by kprakasam** [comment]
@yangminzhu I sent PR on June 1,  before your proposal, if you look at the proposal it is exactly the same, unified matching API and implementation could be a library target shared by waf, rbac, etc.

**Comment by yangminzhu** [comment]
> @yangminzhu I sent PR on June 1, before your proposal, 

It's not about who send out the PR/proposal first, I didn't know this until noticed by Harvey. Both the matching API here and the one in my proposal is just based on the RBAC API because it's already used in the control plane (istio) since we originally introduced it. Let's focus on having a better and consistent matching API in Envoy instead of arguing who sent out the PR earlier.

> if you look at the proposal it is exactly the same, unified matching API and implementation could be a library target shared by waf, rbac, etc.

It's not the same and actually quite different. The intention is to create a common shared matching API that can be used in different filters, instead of embedding it in any of the specific filter (i.e. `filters/http/waf/v3/waf.proto`).  And from what I have seen in the PR, it looks like the waf filter doesn't need anything special from the matching API, it can just refer to the unified matching API where is likely to be placed in a common directory. This is the things that we should consolidate and share, does this look good to you?

**Comment by kprakasam** [comment]
Let me rephrase I meant to say that I was not aware when I sent the PR of your proposal. I can use your matching API, one thing missing matcher is a CIDR set, which will be useful when blocking multiple CIDR ranges and more efficient.

**Comment by yangminzhu** [comment]
Thanks, that proposal is still pending for feedback and will have some changes to make sure it support all different use cases (e.g. a CIDR set). Once @mattklein123 is back from vacation, I think it will get some more feedback and we can discuss it further in the next Envoy community meeting, wdyt?

**Comment by kprakasam** [comment]
Sure no problem.
